plugins/wasmedge_zlib/zlibfunc.cpp

```diff
@@ -0,0 +1,185 @@
+  if (WasmZlibVersion[0] != ZLIB_VERSION[0]) {
+    spdlog::error("[WasmEdge Zlib] [WasmEdgeZlibDeflateInit_] Major Zlib "
+                  "version of Host ({}) & Wasm ({}) "
+                  "does not match",
+                  ZLIB_VERSION[0], WasmZlibVersion[0]);
+    return Unexpect(ErrCode::Value::HostFuncError);
+  }
+
+  auto HostZStream = std::make_unique<z_stream>();
+
+  HostZStream.get()->zalloc = Z_NULL;
+  HostZStream.get()->zfree = Z_NULL;
+  HostZStream.get()->opaque =
+      Z_NULL; // ignore opaque since zmalloc and zfree was ignored
+
+  const auto z_res =
+      deflateInit_(HostZStream.get(), Level, ZLIB_VERSION, sizeof(z_stream));
+
+  Env.ZStreamMap.emplace(std::make_pair(ZStreamPtr, std::move(HostZStream)));
+
+  return static_cast<int32_t>(z_res);
+}
+
+Expect<int32_t>
+WasmEdgeZlibInflateInit_::body(const Runtime::CallingFrame &Frame,
+                               uint32_t ZStreamPtr, uint32_t VersionPtr,
+                               int32_t StreamSize) {
+
+  auto *MemInst = Frame.getMemoryByIndex(0);
+  if (MemInst == nullptr) {
+    return Unexpect(ErrCode::Value::HostFuncError);
+  }
+  // Wasm_z_stream *WasmZStream = MemInst->getPointer<Wasm_z_stream
+  // *>(ZStreamPtr);
+  if (StreamSize != 56) {
+    spdlog::error("[WasmEdge Zlib] [WasmEdgeZlibInflateInit_] WASM "
+                  "sizeof(z_stream) != 56 but {}",
+                  StreamSize);
+    return Unexpect(ErrCode::Value::HostFuncError);
+  }
+
+  const char *WasmZlibVersion = MemInst->getPointer<const char *>(VersionPtr);
+  if (WasmZlibVersion[0] != ZLIB_VERSION[0]) {
+    spdlog::error("[WasmEdge Zlib] [WasmEdgeZlibInflateInit_] Major Zlib "
+                  "version of Host ({}) & Wasm ({}) "
+                  "does not match",
+                  ZLIB_VERSION[0], WasmZlibVersion[0]);
+    return Unexpect(ErrCode::Value::HostFuncError);
+  }
+
+  auto HostZStream = std::make_unique<z_stream>();
+
+  HostZStream.get()->zalloc = Z_NULL;
+  HostZStream.get()->zfree = Z_NULL;
+  HostZStream.get()->opaque =
+      Z_NULL; // ignore opaque since zmalloc and zfree was ignored
+
+  const auto z_res =
+      inflateInit_(HostZStream.get(), ZLIB_VERSION, sizeof(z_stream));
+
+  Env.ZStreamMap.emplace(std::make_pair(ZStreamPtr, std::move(HostZStream)));
+
+  return static_cast<int32_t>(z_res);
+}
+
+Expect<int32_t> WasmEdgeZlibDeflate::WasmEdgeZlibDeflate::body(
+    const Runtime::CallingFrame &Frame, uint32_t ZStreamPtr, int32_t Flush) {
+
+  const auto HostZStreamIt = Env.ZStreamMap.find(ZStreamPtr);
+  if (HostZStreamIt == Env.ZStreamMap.end()) {
+    return Unexpect(ErrCode::Value::HostFuncError);
+  }
+  auto HostZStream = HostZStreamIt->second.get();
+  auto *MemInst = Frame.getMemoryByIndex(0);
+  Wasm_z_stream *WasmZStream = MemInst->getPointer<Wasm_z_stream *>(ZStreamPtr);
+  // get the whole memory from start
+  unsigned char *WasmMemStart = MemInst->getPointer<unsigned char *>(0);
```

Comment:
Am I mistaken?

Shouldn't `MemInst->getPointer<unsigned char *>(0);` just hand me the whole `wasm linear memory` base address?

That's what I want here, and the I can perform pointer arithmetic on them.

Reply:
The `wasm linear memory` starts from 0. However, the application will also put data in memory. How do you ensure no other part uses memory before the zlib section?

Reply:
I don't care what the application puts there, I just need the `base address`.
After that I will also have the required `Offset` where I actually need to perform stuff, and where zlib related things reside.
I will just use the `base address` + `Offset` to translate the wasm memory address to the host address.

Reply:
You should never get the raw pointer from MemInst and store it. The raw pointer may change once the memory grows. And it's not possible to use the native pointer because your application can only see the Wasm level. You should always use offset to get pointers from the MemInst.

Reply:
Okay, I will change it.

Reply:
Can I guareentee that the `raw pointer` will be valid in the execution duration of that particular host function?

Reply:
The pointer will keep the same in the host function. Because you cannot know if other instructions will trigger `memory.grow` outside of the host function itself.

Reply:
I mean't, if I store the `raw pointer` for only that particular execution of that host function, shouldn't that be safe?